### ISSUE TITLE
test_driver actions don't work on targets in shadow DOM

### ISSUE THREAD

**Issue by nt1m** [web-platform-tests/wpt#41257]
After merging https://github.com/web-platform-tests/wpt/pull/41251 (which works fine on webkit's test runner infrastructure), I realized test_driver does not actually support targets in shadow DOM, because it tries to create a selector, which doesn't really work in a shadow dom context.

cc @domenic @mfreed7 @josepharhar @rniwa @gsnedders @jgraham

**Comment by mfreed7** [comment]
Can you point to where it's trying to create a selector from `element`? I feel like I've used `test_driver.Actions().pointerMove(x,y,{origin: element})` with `element` in a shadow root and things worked correctly. But I could be mistaken. I do notice none of the popover/shadow tests use it.

**Comment by nt1m** [comment]
https://github.com/web-platform-tests/wpt/blob/109ee2d869715b4338e78e3369a9fa59550974f1/tools/wptrunner/wptrunner/testdriver-extra.js#L220

You can see the test failing here in Chrome for this exact reason: https://wpt.fyi/results/html/semantics/popovers/popover-light-dismiss-flat-tree.html?diff&filter=ADC&run_id=5102411146854400&run_id=5205993141567488

```
ValueError: Selector '#\69 \6e \6e \65 \72 \50 \6f \70 \6f \76 \65 \72 \54 \6f \67 \67 \6c \65 ' matches no elements
```

**Comment by domenic** [comment]
I feel like in the past there was talk of adding shadow-piercing selectors specifically for use in automation tooling, not exposed to the web at large. I don't know where those discussions happened or where they ended up though...

I'll blindly tag @mathiasbynens as my general "person who knows these sorts of things".

**Comment by mathiasbynens** [comment]
We recently revisited this in the context of WebDriver BiDi: https://github.com/w3c/webdriver-bidi/issues/342 The outcome was that at this point, this functionality should be implemented by clients rather than built into the spec.

**Comment by jgraham** [comment]
Right, I think the WebDriver group in particular don't want to try to specify custom selectors. If the CSSWG wanted to specify it they would probably be used.

In the meantime this is fixable with some effort; if you can get a reference to the element with script then at least in theory we can split the element selection into multiple parts, one for each shadow tree. Then the WebDriver side should be able to do the shadow tree traversal. Alternatively one could make it work for the same-origin case by not depending on `postMessage` for communication with the top-level test window and directly sending the element back rather than a selector that should match the element.

We're probably also pretty close to the point where we could implement a WebDriver-BiDi backend for wptrunner / testdriver, which would solve this problem by allowing us to avoid the pile of hacks required to handle the fact that WebDriver requires synchronous, blocking communication with a single Window at a time.

**Comment by josepharhar** [comment]
test_driver.click() doesn't work on elements in shadowdom, which I tried to fix here but it got complicated and I gave up for now: https://github.com/web-platform-tests/wpt/pull/37483
I don't think it was due to an issue with selectors, but it does fit under the title of this issue

**Comment by gsnedders** [comment]
> test_driver.click() doesn't work on elements in shadowdom, which I tried to fix here but it got complicated and I gave up for now: #37483 I don't think it was due to an issue with selectors, but it does fit under the title of this issue

See also #37942

**Comment by emilio** [comment]
I have a workaround for this for the popover tests in https://phabricator.services.mozilla.com/D203152.
